tests/test_optimize_shot_adapative.py

```diff
@@ -0,0 +1,514 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for the shot adaptive optimizer"""
+import pytest
+from scipy.stats import multinomial
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+class TestExceptions:
+    """Test exceptions are raised for incorrect usage"""
+
+    def test_analytic_device_error(self):
+        """Test that an exception is raised if an analytic device is used"""
+        H = qml.Hamiltonian([0.3, 0.1], [qml.PauliX(0), qml.PauliZ(0)])
+        dev = qml.device("default.qubit", wires=1, shots=None)
+        expval_cost = qml.ExpvalCost(lambda x, **kwargs: qml.RX(x, wires=0), H, dev)
+
+        opt = qml.ShotAdaptiveOptimizer(min_shots=10)
+
+        # test expval cost
+        with pytest.raises(ValueError, match="can only be used with devices that"):
+            opt.step(expval_cost, 0.5)
+
+        # test qnode cost
+        with pytest.raises(ValueError, match="can only be used with devices that"):
+            opt.step(expval_cost.qnodes[0], 0.5)
+

```

Comment:
Don't forget to remove this as well as the method that this is testing, since the shots refactor is already merged!

Reply:
Would any other changes be needed due to this? 🤔

Reply:
I ended up having to keep this 🤔 I just changed the exception message to remove mention of the `analytic` keyword arg.

The reason being:

After shots II:

- If a user creates a QNode with shots=None, the QNode will choose default.qubit.autograd.
- If a user creates a QNode with shots>0, the QNode will choose default.qubit

Which is :ok_hand:

However, if a user creates a QNode with shots=None , and then later changes the number of shots, they will run into problems, because default.qubit.autograd does not like random sampling.

Alternatively, if a user creates a QNode with shots>0, and then later changes to shots=None, they will be stuck with the parameter-shift rule

So two thoughts:

- We either leave it as is, and in the documentation make it clear you can't switch between shots=None and shots>0 on the fly
- Or, we have the QNode re-do the device picking every time shots changes

With respect to this PR, I thought you would be able to do this:
```python
dev = qml.device("default.qubit", wires=2)
@qml.qnode(dev)
    def circuit(...)
params = opt_adaptive_shot.step(circuit, params)
```
However, that last line fails for the reasons above; it works if you change the device line to
```python
dev = qml.device("default.qubit", wires=2, shots=100)
```